text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,454 @@
+```ts
+import * as cdkp from '@aws-cdk/pipelines';
+
+const pipeline = new cdkp.Pipeline(this, 'Pipeline', {
+  build: new cdkp.Build({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: cdkp.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+  backend: new cdkp.CodePipelineBackend(),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+pipeline.addStage(stage, {
+  after: [
+    new cdkp.ShellScriptAction({
+      input: pipeline.build.additionalOutput('tests'),
+      commands: ['node ./integ-tests'],
+    }),
+  ],
+});
+```
+
+### Splitting definition and backend
+
+It is helpful if there is a clearly defined moment in time when the pipeline definition is ready to be rendered.
+
+There are various ways to organize this, each pretty much isomorphic to one another.
+
+OPTION 1
+
+Clear split of definition and instantiation. Very much like StepFunctions.
+
+```ts
+const pipelineDefinition = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+});
+pipelineDefinition.addCdkStage(...);
+pipelineDefinition.addCdkStage(...);
+
+new cdkp.CodePipelineBackend(this, 'Pipeline', {
+  pipelineDefinition: pipelineDefinition,
+});
+```
+
+OPTION 2
+
+Automatic rendering moment.
+
+```ts
+const pipeline = new cdkp.Pipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+  backend: new cdkp.CodePipelineBackend({
+    ...backendOptions
+}),
+});
+pipeline.addCdkStage(...);
+pipeline.addCdkStage(...);
+
+pipeline.render();  // <-- can be automatically called if not called at `prepare()` time.
+```
+
+OPTION 3 (syntactic sugar over OPTIONS 1 or 2)
+
+Backend-specific syntactic sugar for people that definitely only need one backend.
+
+```ts
+const pipeline = new cdkp.CodePipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    ...
+  }),
+  ...backendOptions
+});
+pipeline.addCdkStage(...);
+pipeline.addCdkStage(...);
+
+pipeline.render();  // <-- can be automatically called if not called at `prepare()` time.
+```
+
+Advantages of options 2, 3 are that `render()` can be implicit if necessary, and so cannot
+be accidentally forgotten.
+
+Recommendation: we implement options 1 and 3.
+
+### Waves
+
+Deploying multiple applications in parallel is done by adding waves:
+
+```ts
+const def = new cdkp.PipelineDefinition({
+  build: new cdkp.CdkBuild({ ... }),
+  waves: [
+    {
+      waveName: '...',
+      stages: [
+        {
+          stageName?: '...',
+          stage: new MyAppStage(...),
+          after: [...],
+        }
+      ],
+    },
+  ],
+});
+def.addCdkStage(new MyAppStage(...));
+
+const wave = def.addWave('asdf', {
+  stages: [...],
+});
+wave.addCdkStage(new MyAppStage(...));
```

Comment:
I would like to comment on the use of `stage` terminology as part of waves. From my experience, customers are familiar with `deployment` term (especially from Kubernetes world), but not so much with `stage`. I understand we probably can’t change `cdk.Stage` to `cdk.Deployment`, but can we at least change the CDK Pipelines interface from `wave.addCdkStage` to `wave.addDeployment`? :slightly_smiling_face:. Example to illustrate below.

```
pipeline = pipelines.Pipeline()

europe_wave = pipeline.add_wave('Europe')

eu_west_1_env = cdk.Environment(account='111111111111', region='eu-west-1')
# class Deployment(cdk.Stage): ...
eu_west_1_deployment = Deployment(self, APPLICATION_NAME, env=eu_west_1_env)
eu_west_1_smoke_test = pipelines.ShellScriptAction(
    action_name='SmokeTest', commands=['curl https://eu-west-1.my.webservice.com/'])
europe_wave.add_deployment(eu_west_1_deployment, after=[eu_west_1_smoke_test])

eu_west_2_env = cdk.Environment(account='222222222222', region='eu-west-2')
# class Deployment(cdk.Stage): ...
eu_west_2_deployment = Deployment(self, APPLICATION_NAME, env=eu_west_2_env)
eu_west_2_smoke_test = pipelines.ShellScriptAction(
    action_name='SmokeTest', commands=['curl https://eu-west-2.my.webservice.com/'])
europe_wave.add_deployment(eu_west_2_deployment, after=[eu_west_2_smoke_test])
```